### ISSUE TITLE
Add css style conditions

### ISSUE THREAD

**Issue by mseele** [codeborne/selenide#569]
Add css style check conditions into selenide. 

We use this own written helper class until selenide supports it:

```
import static java.util.Arrays.*;
import static java.util.Collections.*;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.StringJoiner;

import org.openqa.selenium.WebElement;

import com.codeborne.selenide.CollectionCondition;
import com.codeborne.selenide.Condition;
import com.codeborne.selenide.ex.ElementNotFound;
import com.codeborne.selenide.ex.UIAssertionError;
import com.codeborne.selenide.impl.WebElementsCollection;

public class Styles extends CollectionCondition {

	public static Condition style(final Style expectedStyle) {
		return new Condition("style") { //$NON-NLS-1$
			@Override
			public boolean apply(WebElement element) {
				return Styles.apply(element, expectedStyle);
			}

			@Override
			public String actualValue(WebElement element) {
				return Styles.actualValue(element, expectedStyle);
			}

			@Override
			public String toString() {
				return "style " + Styles.expectedValue(expectedStyle); //$NON-NLS-1$
			}
		};
	}

	public static CollectionCondition styles(Style... expectedStyles) {
		return new Styles(expectedStyles);
	}

	protected final List<Style> fExpectedStyles;

	private Styles(Style... expectedStyles) {
		this(asList(expectedStyles));
	}

	private Styles(List<Style> expectedStyles) {
		if (expectedStyles.isEmpty()) {
			throw new IllegalArgumentException("No expected styles given"); //$NON-NLS-1$
		}
		this.fExpectedStyles = unmodifiableList(expectedStyles);
	}

	@Override
	public boolean apply(List<WebElement> elements) {
		if (elements.size() != this.fExpectedStyles.size()) {
			return false;
		}

		for (int i = 0; i < this.fExpectedStyles.size(); i++) {
			if (!apply(elements.get(i), this.fExpectedStyles.get(i))) {
				return false;
			}
		}
		return true;
	}

	@Override
	public void fail(WebElementsCollection collection, List<WebElement> elements, Exception lastError, long timeoutMs) {
		if (elements == null || elements.isEmpty()) {
			ElementNotFound elementNotFound = new ElementNotFound(collection, expectedValues(), lastError);
			elementNotFound.timeoutMs = timeoutMs;
			throw elementNotFound;
		}
		throw new StyleMismatch(collection, actualValues(elements), expectedValues(), timeoutMs);
	}

	private List<String> actualValues(List<WebElement> elements) {
		List<String> values = new ArrayList<>();
		for (WebElement element : elements) {
			int index = elements.indexOf(element);
			if (index > 0 && index < this.fExpectedStyles.size()) {
				values.add(actualValue(element, this.fExpectedStyles.get(index)));
			}
		}
		return values;
	}

	private List<String> expectedValues() {
		List<String> values = new ArrayList<>();
		for (Style expectedStyle : this.fExpectedStyles) {
			values.add(expectedValue(expectedStyle));
		}
		return values;
	}

	static boolean apply(WebElement element, Style expectedStyle) {
		for (Entry<String, String> entry : expectedStyle.cssValues().entrySet()) {
			if (!entry.getValue().equals(element.getCssValue(entry.getKey()))) {
				return false;
			}
		}
		return true;
	}

	static String actualValue(WebElement element, Style expectedStyle) {
		StringJoiner styles = new StringJoiner(" + "); //$NON-NLS-1$
		for (Entry<String, String> entry : expectedStyle.cssValues().entrySet()) {
			String key = entry.getKey();
			styles.add(key + ": " + element.getCssValue(key)); //$NON-NLS-1$
		}
		return styles.toString();
	}

	static String expectedValue(Style expectedStyle) {
		StringJoiner styles = new StringJoiner(" + "); //$NON-NLS-1$
		for (Entry<String, String> entry : expectedStyle.cssValues().entrySet()) {
			styles.add(entry.getKey() + ": " + entry.getValue()); //$NON-NLS-1$
		}
		return styles.toString();
	}

	@Override
	public String toString() {
		return "Styles " + this.fExpectedStyles; //$NON-NLS-1$
	}

	public static enum Style {
		Effective("font-style: normal", "color: rgba(51, 51, 51, 1)"), //$NON-NLS-1$ //$NON-NLS-2$
		Proposal("font-style: italic", "color: rgba(51, 51, 51, 1)"), //$NON-NLS-1$ //$NON-NLS-2$
		ProposalFallback("font-style: italic", "color: rgba(185, 185, 185, 1)"), //$NON-NLS-1$ //$NON-NLS-2$

		TreeItemEffective("font-style: normal"), //$NON-NLS-1$
		TreeItemProposal("font-style: italic"), //$NON-NLS-1$

		Selected("border-bottom-style: solid", "border-bottom-color: rgba(11, 94, 215, 1)"), //$NON-NLS-1$ //$NON-NLS-2$
		NotSelected("border-bottom-style: none"), //$NON-NLS-1$

		Approved("font-style: normal", "color: rgba(51, 51, 51, 1)"), //$NON-NLS-1$ //$NON-NLS-2$
		NotApproved("font-style: italic", "color: rgba(93, 93, 93, 1)"), //$NON-NLS-1$ //$NON-NLS-2$

		InfoNotification("background-color: rgba(255, 229, 184, 1)"), //$NON-NLS-1$
		ErrorNotification("background-color: rgba(255, 205, 210, 1)"); //$NON-NLS-1$

		private Map<String, String> fCssValues;

		private Style(String...cssValues) {
			this.fCssValues = new HashMap<>();
			for (String cssValue : cssValues) {
				String key = cssValue.substring(0, cssValue.indexOf(':')).trim();
				String value = cssValue.substring(cssValue.indexOf(':') + 1).trim();
				this.fCssValues.put(key, value);
			}
		}

		public Map<String, String> cssValues() {
			return this.fCssValues;
		}

	}

	public static class StyleMismatch extends UIAssertionError {

		public StyleMismatch(WebElementsCollection collection, List<String> actualStyles, List<String> expectedStyles, long timeoutMs) {
			super("\nActual:   " + actualStyles + "\nExpected: " + expectedStyles + "\nCollection: " + collection.description()); //$NON-NLS-1$ //$NON-NLS-2$ //$NON-NLS-3$
			super.timeoutMs = timeoutMs;
		}

		@Override
		public String toString() {
			return getClass().getSimpleName() + ' ' + getMessage() + uiDetails();
		}
	}

}

```

**Comment by asolntsev** [comment]
@mseele I am not sure it's a good idea to check style. 
The reasoning is that **functional** tests should check functionality, not the look & feel. These are things that robot cannot test well. Robot can check that element has `background-color: rgba(255, 229, 184, 1)`, but it cannot check if it's nice or ugly. 

Additionally, it makes tests unstable. Every time when designer changes some fonts, colors or sizes, your tests will fail. 

What I recommend is to check **CSS class** in tests, like:

`$("#alert").shouldHave(cssClass("info-notification"));`

It tests the business value: that the given text is **info notification**. And how it looks like is **implementation detail**. It should be describe in CSS and checked by human.

**Comment by mseele** [comment]
@asolntsev You're wright that the best case would be to test the css class.
We are developing with gwt and using CssResource (http://www.gwtproject.org/doc/latest/DevGuideClientBundle.html#CssResource) which uses obfuscation. So we don't know the css class name, it's different after each compile.

This is why we use specific style attributes to detect function (like selected, info notification etc.). If this attributes will be changed, we need to change the test attributes. There is no other chance for us to detect these functionality.

**Comment by asolntsev** [comment]
@mseele Hm... Probably it's valid case for adding `shouldHave(style())` condition.

Thought, as I see from provided GWT documentation, you can add custom IDs and classes in GWT:

> All of the widgets created with the GWT toolkit will have a default class name, but a widget’s style name can be set using `setStyleName()`. Static elements can have their class set in the HTML source code for your application.

By default, neither the browser nor GWT creates default id attributes for widgets. You must explicitly create an id for the elements you want to refer to in this manner, and you must insure that each “id” value is unique. A common way to do this is to set them on static elements in your HTML host page:

```
<div id="my-button-id"/>
```

To set the id for a GWT widget, retrieve its DOM Element and then set the id attribute as follows:

```
Button b = new Button();
  DOM.setElementAttribute(b.getElement(), "id", "my-button-id")
```

This would allow you to reference a specific widget in a style sheet as follows:

```
#my-button-id { font-size: 100%; }
```

**Comment by asolntsev** [comment]
It seems that this issue is duplicate of https://github.com/codeborne/selenide/issues/628 which was implemented in Selenide 4.11.3: http://selenide.org/2018/05/08/selenide-4.11.3/

@mseele Can you confirm?
